batch/sql/estimated-current.sql

```diff
@@ -1306,6 +1338,85 @@ BEGIN
+                              FOR UPDATE);
+
+  IF cur_batch_state = 'running' AND NOT cur_cancelled THEN
+    INSERT INTO user_inst_coll_resources (user, inst_coll, token,
+      n_ready_jobs, ready_cores_mcpu,
+      n_running_jobs, running_cores_mcpu,
+      n_creating_jobs,
+      n_cancelled_ready_jobs, n_cancelled_running_jobs, n_cancelled_creating_jobs)
+    SELECT user, inst_coll, 0,
+      -1 * (@n_ready_cancellable_jobs := COALESCE(SUM(n_ready_cancellable_jobs), 0)),
+      -1 * (@ready_cancellable_cores_mcpu := COALESCE(SUM(ready_cancellable_cores_mcpu), 0)),
+      -1 * (@n_running_cancellable_jobs := COALESCE(SUM(n_running_cancellable_jobs), 0)),
+      -1 * (@running_cancellable_cores_mcpu := COALESCE(SUM(running_cancellable_cores_mcpu), 0)),
+      -1 * (@n_creating_cancellable_jobs := COALESCE(SUM(n_creating_cancellable_jobs), 0)),
+      COALESCE(SUM(n_ready_cancellable_jobs), 0),
+      COALESCE(SUM(n_running_cancellable_jobs), 0),
+      COALESCE(SUM(n_creating_cancellable_jobs), 0)
+    FROM job_group_inst_coll_cancellable_resources
+    JOIN batches ON batches.id = job_group_inst_coll_cancellable_resources.batch_id
+    INNER JOIN batch_updates ON job_group_inst_coll_cancellable_resources.batch_id = batch_updates.batch_id AND
+      job_group_inst_coll_cancellable_resources.update_id = batch_updates.update_id
+    WHERE job_group_inst_coll_cancellable_resources.batch_id = in_batch_id AND
+      job_group_inst_coll_cancellable_resources.job_group_id = in_job_group_id AND
+      batch_updates.committed
+    GROUP BY user, inst_coll
+    ON DUPLICATE KEY UPDATE
+      n_ready_jobs = n_ready_jobs - @n_ready_cancellable_jobs,
+      ready_cores_mcpu = ready_cores_mcpu - @ready_cancellable_cores_mcpu,
+      n_running_jobs = n_running_jobs - @n_running_cancellable_jobs,
+      running_cores_mcpu = running_cores_mcpu - @running_cancellable_cores_mcpu,
+      n_creating_jobs = n_creating_jobs - @n_creating_cancellable_jobs,
+      n_cancelled_ready_jobs = n_cancelled_ready_jobs + @n_ready_cancellable_jobs,
+      n_cancelled_running_jobs = n_cancelled_running_jobs + @n_running_cancellable_jobs,
+      n_cancelled_creating_jobs = n_cancelled_creating_jobs + @n_creating_cancellable_jobs;
+
+    # delete all rows that are children of this job group
+    DELETE job_group_inst_coll_cancellable_resources FROM job_group_inst_coll_cancellable_resources
+    LEFT JOIN batch_updates ON job_group_inst_coll_cancellable_resources.batch_id = batch_updates.batch_id AND
+      job_group_inst_coll_cancellable_resources.update_id = batch_updates.update_id
+    INNER JOIN job_group_self_and_ancestors ON job_group_inst_coll_cancellable_resources.batch_id = job_group_self_and_ancestors.batch_id AND
+      job_group_inst_coll_cancellable_resources.job_group_id = job_group_self_and_ancestors.job_group_id
+    WHERE job_group_inst_coll_cancellable_resources.batch_id = in_batch_id AND
+      job_group_self_and_ancestors.ancestor_id = in_job_group_id AND
```

Comment:
Hmm. Looking at this use case I wonder if a better name for this is `job_group_transitive_dag` and the two fields should be `ancestor_and_self_id` `descendent_and_self_id`. In particular,
```
select descendent_and_self_id
from job_group_transitive_dag
where ancestor_and_self_id = 123 and batch_id = 1
```
Gives me all the descendants of job group 123 (and itself), right? Contrariwise, this gives me all the ancestors of 123 (and itself).
```
select ancestor_and_self_id
from job_group_transitive_dag
where descendent_and_self_id = 123 and batch_id = 1
```
These nicely correspond to the [up-set and down-set](https://en.wikipedia.org/wiki/Upper_set) from order theory.

Reply:
Maybe `above_id` and `below_id` are succinct and accurate? Those remove the non-reflexive connotation of "ancestor".

Reply:
I can try that, but I think we'd need to potentially do an offline migration for the rename unless we're okay with not being able to create new batches until the new batch code is deployed after the migration.

Reply:
Let's punt on the name for now. I think we're getting closer to something reasonable, but no need to disrupt this PR over it.